Index: .gitignore
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+># Created by .ignore support plugin (hsz.mobi)\n### Example user template template\n### Example user template\n\n# IntelliJ project files\n.idea\n*.iml\nout\ngen\nsrc/nanocompare/.DS_Store\n/figures/\n/result/\nwork\n.nextflow*\nnextflow\n.ipynb_checkpoints\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.gitignore b/.gitignore
--- a/.gitignore	(revision 57f724d0f7b245427898dc7758857009b73782a0)
+++ b/.gitignore	(date 1619889249850)
@@ -3,8 +3,8 @@
 ### Example user template
 
 # IntelliJ project files
-.idea
-*.iml
+#.idea
+#*.iml
 out
 gen
 src/nanocompare/.DS_Store
